utils/zerovec/design_doc.md

```diff
@@ -0,0 +1,426 @@
+ 
+The latter two features may potentially be toggled via attributes.
+
+`#[make_varule]` will know the "corresponding unsized type" for various fixed-size heap types:
+
+ - `String` maps to `str`
+ - `[u8]`, `Box<[u8]>`, `Vec<u8>` maps to `[u8]`
+ - `Vec<T>`, `Box<[T]>`, `ZeroVec<T>`, `&ZeroSlice<T>` map to `ZeroSlice<T>`
+ - `Vec<T>` where `T` is `Vec`, `String`, or a `ZeroVec` type maps to `VarZeroSlice<T>`
+ - `VarZeroVec<T>`, `VarZeroVecOwned<T>`, `&VarZeroSlice<T>` all map to `VarZeroSlice<T>`
+ 
+
+It can either require this type to be at the end of the struct, or magically figure out which it is by looking at the types and move it to the end.
+
+When applied to a struct with a single heap field, `#[make_varule]` will behave largely similar to `#[make_ule]`, generating a `#[derive(VarULE)]` type with a corresponding `F::ULE` field for each field in `Foo`, and with a final unsized field of the "corresponding unsized type" of the fixed size heap type. `EncodeAsVarULE` just needs to encode each field in order, and that should be it.
+
+For structs with multiple heap fields, `#[make_varule]` can generate an additional private `length` field for each dynamically sized field, with a single `data: [u8]` buffer at the end, and convenience getters for the fields.
+
+
+For enums with heap fields, `#[make_varule]` can apply a similar scheme as `#[make_ule]`, discussed below.
+
+For an initial pass, we may only support single-heap-field structs for `#[make_varule]`.
+
+
+### Handling enums
+
+Ideally, enums can have ULE impls autogenerated for them, but handling the discriminants gets tricky.
+
+One way to handle this is to generate a private internal struct for each variant and do the usual ULE or VarULE generation for each. Then, manaully construct a `#[repr(packed)]` tagged union with a `u8` tag and a `union` of all of the internal structs. This is somewhat complicated but actually relatively simple to implement since it can use `#[make_ule]` and `#[make_varule]`.
+
+
+We may also do this completely manually, which gives us the opportunity for bitpacking the discriminant further, if combined with the bitpacking scheme discussed below.
+
+### Bitpacking
+
+The proc macro as designed so far will take up multiples of eight bits for each field. However, many fields (like a hypothetical `BoolULE`, or `ULE` types for dataless enums) need significantly less space. It would be nice if we could get the proc macro to automatically perform bitpacking.
+
+The way this can be achieved is the following.
+
+Firstly, `ULE` gets `BITS` and `BYTES` associated constants. `BYTES` is always `sizeof::<Self>()` (but as an associated constant so it may be used in generic const contexts). `BITS` by default is just `BYTES * 8`, however in some cases it may be smaller.
```

Comment:
```suggestion
Firstly, `ULE` gets `BITS` and `BYTES` associated constants. `BYTES` is always `sizeof::<Self>()` (but as an associated constant so it may be used in generic const contexts). `BITS` by default is just `BYTES * 8`; however, in some cases it may be smaller.
```